### DISCUSSION TITLE
Ideas around disposing providers on log out

### DISCUSSION THREAD

**Discussion by chimon2000** [rrousselGit/riverpod#3023] [Q&A]
If you have any number of providers that need to be disposed of at logout so that a new user doesn't see cached values, how are you handling that scenario? Are you just manually tracking providers?

**Comment by AhmedLSayed9** [comment]
1. I prefer usually using autoDispose providers when possible.
2. For keep-alive or dispose-delayed providers, I let them watch authState provider. so, when auth state changes from login to logout, they will be disposed.

**Reply by AhmedLSayed9** [reply]
number 2 is easily done by defining an extension:
```dart
extension RefExtension on Ref {
  void watchAuthState() {
    watch(authStateProvider.select((user) => user != null));
  }
}

//At the start of the provider body, I just add:
ref.watchAuthState();
```
